Add explicit_columns_select ActiveRecord configuration option

maciek-rr opened the pull request:
### Summary

Column-adding migrations have the potential to fail transactions that are happening alongside, when queries are cached as prepared statements in PostgreSQL.
This is caused by result type of prepared statement being changed mid-transaction, PostgreSQL signals error which gets wrapped in the `ActiveRecord::PreparedStatementCacheExpired` exception.

This PR introduces a new (disabled by default) setting for ActiveRecord `explicit_columns_select`, it can be used both globally and on the individual models' level. 

When enabled, AR queries use explicit colum names in selects and not `*`. 
Such queries remain unaffected by migrations adding new columns.

If this solution is acceptable and makes sense to be included in Rails, I'll add the changelog entry.

### Other Information

#### Non-function considerations:
- Performance of both types of queries is understood to be the same
- Listing all column names in the select clause adds noticeable overhead to pathological cases of models with large number of columns

#### Currently existing workarounds:
- Disabling prepared statements altogether - this has a significant negative impact on the database load
- Setting `self.ignored_columns = %w[nonexistent_column]` - triggers the same code path as this patch, but as every hack, requires comment in each place it's applied, also needs to be set individually on each affected model


<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

maciek-rr committed:
Add explicit_columns_select ActiveRecord configuration option

Column adding migrations have the potential to fail transactions
happening in parallel, when queries are cached using prepared
statements in PostgreSQL.

This is caused by result type being changed mid-transaction.

This PR introduces a new setting for ActiveRecord `explicit_columns_select`.
When enabled, AR queries by default use explicit colum names in selects
and not star, such queries remain unaffected by migrations adding new columns.
The setting can be used globally and on the individual model level.

simi commented:
:thinking: Even this seems to be small change code-wise, I think it could be potentially hard to maintain.

Can't you rescue `ActiveRecord::PreparedStatementCacheExpired` and handle on your own in your app?

irvingwashington commented:
Hey, thanks for quick feedback. 
It's the same behavior as we have when excluding columns on the model level, there's no new logic.
The problem with handling it the way you describe is that you'd have to add a rescue clause to each transaction in your codebase and while it's doable and viable for simple cases (just retry), for more complex cases (nested transactions) it can actually be very tricky and error prone.

maciek-rr committed:
Fix the explicit_columns_select flag test

- make it db engine agnostict
- make sure it doesn't leak the configuration setting change

simi commented:
Yeah, I do follow. But this seems like a huge change to fix `ActiveRecord::PreparedStatementCacheExpired` which is edge case IMHO.

If I understand it well, cache is cleared once this is raised:
https://github.com/rails/rails/blob/31be40d1dcf16e33c862701652eb735d13b1b3f5/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L342-L346

Isn't there any other possible solution for this? I have seen recently it is getting popular to inherit database adapter and do those tweaks in there. You should be able to retry on problem in there. Would retrying with cleared cache work for you?

PS: Is there any idea why ActiveRecord doesn't handle this problem on its own?

irvingwashington commented:
I agree, this is a big change to partially address a PostgreSQL corner case.
The ideal solution would work on the Postgres adapter level and would just transform star queries into explicit column selects before creating prepared statements, but I doubt it's doable without the whole context from ActiveRecord.
For now I guess I'll work on it in a separate gem.

simi commented:
@irvingwashington I'm basically just randomly passing by not affiliated with Rails core team at all. It would be better to wait for someone from Rails core team for proper review.

rishav-csenitjsr commented:
@irvingwashington Can you squash all commits in one, please?

eugeneius commented:
Let's figure out if the feature will be accepted first; otherwise getting the patch ship-shape would be wasted effort.

I think it's worth asking: why does PostgreSQL return this error? Why can't it reuse the prepared statement after a new column has been added?

Here's where the error is reported:

https://github.com/postgres/postgres/blob/b150a76793109b00ea43c26e0006b3cad9030096/src/backend/utils/cache/plancache.c#L716-L718

It's inside a conditional that checks `equalTupleDescs()`:

https://github.com/postgres/postgres/blob/b150a76793109b00ea43c26e0006b3cad9030096/src/backend/utils/cache/plancache.c#L712

A comment a few lines up says "should we use a weaker condition than equalTupleDescs() here?":

https://github.com/postgres/postgres/blob/b150a76793109b00ea43c26e0006b3cad9030096/src/backend/utils/cache/plancache.c#L700-L701

Grepping for `equalTupleDescs`, I found a similar function related to updating views:

> This is basically a cut-down version of equalTupleDescs(), with code added to generate specific complaints. Also, we allow the new tupledesc to have more columns than the old.

https://github.com/postgres/postgres/blob/b150a76793109b00ea43c26e0006b3cad9030096/src/backend/commands/view.c#L253-L255

It seems like maybe this error is meant to prevent prepared statements from being reused if a column they reference is dropped or changed, but it currently also invalidates statements that use `SELECT *` when a column is added to a referenced table, even though the statement could still be safely reused in that particular case.

If I'm right about that (which is a big if; this is my first time looking at the PostgreSQL codebase), then the ideal solution is to relax the check to allow prepared statements to be reused as long as none of the columns they reference have changed, similar to how `checkViewTupleDesc` works. Every Rails ([and Laravel!](https://github.com/laravel/ideas/issues/1519)) application would get the fix without having to opt in to a new behaviour.

I'm not sure when I'll have time to investigate writing a patch for PostgreSQL to make the above change - if anyone else wants to give it a go, feel free.

simi commented:
@eugeneius Glorious comment! I can try to raise this question at local PostgreSQL group (few PostgreSQL core committers are part of this group) for initial feedback on this.

@irvingwashington do you have any reproducing script around? That will be handy. I can try to prepare some on my own as well.

simi commented:
ok, seems actually really simple to reproduce in psql

```sql
CREATE TABLE test(id int);
PREPARE fooplan(int) AS SELECT * FROM test where id = $1;
ALTER TABLE test ADD COLUMN id2 int;
EXECUTE fooplan(1);
ERROR:  cached plan must not change result type
```

@irvingwashington can you confirm this is minimal script to reproduce your problem?

eugeneius commented:
> I can try to raise this question at local PostgreSQL group (few PostgreSQL core committers are part of this group) for initial feedback on this.

Thanks @simi, I'd appreciate that. 🙂
